File: src/litcoder/module-3/lab-1/MaximumNumberOfSubsequencesInString.py
from typing import List


def solution(text: str, pattern: str) -> int:
    """Returns the maximum no of times pattern can occur as a subsequence of
    modified text

    Args:
        text (str): text
        pattern (str): pattern

    Returns:
        int: max times pattern can  occur as a subsequence
    """
    max_count = 0
    for char in pattern:
        # add char at all positions in text
        for i in range(len(text) + 1):
            new_string = text[0:i] + char + text[i:]

            # generate all subsequence and check count of pattern in it
            subsequences = get_subsequences(new_string)
            max_count = max(max_count, subsequences.count(pattern))

    return max_count


def get_subsequences(string: str) -> List[str]:
    """Returns a list of all possible subsequences of given string

    Args:
        string (str): string

    Returns:
        List[str]: list of all subsequences
    """
    subsequences = [""]

    for char in string:
        curr_subsequence: List[str] = []

        for subsequence in subsequences:
            curr_subsequence.append(subsequence)
            curr_subsequence.append(subsequence + char)

        subsequences = curr_subsequence

    return subsequences


testcases = int(input())
for _ in range(testcases):
    text, pattern = input().split(" ")
    print(solution(text, pattern))

File: codes/college/litcoder/module-3/lab-1/MaximumNumberOfSubsequencesInString.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class MaximumNumberOfSubsequencesInString {

    /**
     * Calculates the maximum number of occurrences of a pattern in a text by adding
     * each character of the pattern at all positions in the text.
     *
     * @param text    the input text
     * @param pattern the pattern to be added and checked for occurrences
     * @return the maximum number of occurrences of the pattern in the text
     */
    public static int solution(String text, String pattern) {
        int maxCount = 0;

        for (char ch : pattern.toCharArray()) {
            // add char at all positions in text
            for (int i = 0; i <= text.length(); i++) {
                StringBuilder newText = new StringBuilder(text);
                newText.insert(i, ch);

                // generate all subsequences and check count of pattern in it
                List<String> subsequences = getSubsequences(newText.toString());
                maxCount = Math.max(maxCount, countOccurrences(subsequences, pattern));
            }
        }

        return maxCount;
    }

    /**
     * Generates all possible subsequences of a given string.
     *
     * @param string the input string
     * @return a list of all subsequences of the input string
     */
    public static List<String> getSubsequences(String string) {
        List<String> subsequences = new ArrayList<>();
        subsequences.add("");

        for (char ch : string.toCharArray()) {
            List<String> currentSubsequence = new ArrayList<>();

            for (String subsequence : subsequences) {
                currentSubsequence.add(subsequence);
                currentSubsequence.add(subsequence + ch);
            }

            subsequences = currentSubsequence;
        }

        return subsequences;
    }

    /**
     * Counts the number of occurrences of a given pattern in a list of
     * subsequences.
     *
     * @param subsequences the list of subsequences to search for the pattern
     * @param pattern      the pattern to count occurrences of
     * @return the number of occurrences of the pattern in the subsequences
     */
    public static int countOccurrences(
            List<String> subsequences,
            String pattern) {
        int count = 0;
        for (String subsequence : subsequences) {
            if (subsequence.equals(pattern)) {
                count++;
            }
        }
        return count;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        String text = scanner.next();
        String pattern = scanner.next();
        System.out.println(solution(text, pattern));

        scanner.close();
    }
}
